File: src/main/java/com/projectswg/holocore/resources/gameplay/world/travel/TravelHelper.java
/***********************************************************************************
 * Copyright (c) 2018 /// Project SWG /// www.projectswg.com                       *
 *                                                                                 *
 * ProjectSWG is the first NGE emulator for Star Wars Galaxies founded on          *
 * July 7th, 2011 after SOE announced the official shutdown of Star Wars Galaxies. *
 * Our goal is to create an emulator which will provide a server for players to    *
 * continue playing a game similar to the one they used to play. We are basing     *
 * it on the final publish of the game prior to end-game events.                   *
 *                                                                                 *
 * This file is part of Holocore.                                                  *
 *                                                                                 *
 * --------------------------------------------------------------------------------*
 *                                                                                 *
 * Holocore is free software: you can redistribute it and/or modify                *
 * it under the terms of the GNU Affero General Public License as                  *
 * published by the Free Software Foundation, either version 3 of the              *
 * License, or (at your option) any later version.                                 *
 *                                                                                 *
 * Holocore is distributed in the hope that it will be useful,                     *
 * but WITHOUT ANY WARRANTY; without even the implied warranty of                  *
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the                   *
 * GNU Affero General Public License for more details.                             *
 *                                                                                 *
 * You should have received a copy of the GNU Affero General Public License        *
 * along with Holocore.  If not, see <http://www.gnu.org/licenses/>.               *
 ***********************************************************************************/
package com.projectswg.holocore.resources.gameplay.world.travel;

import com.projectswg.common.data.encodables.oob.ProsePackage;
import com.projectswg.common.data.encodables.oob.StringId;
import com.projectswg.common.data.location.Location;
import com.projectswg.common.data.location.Terrain;
import com.projectswg.holocore.intents.support.global.chat.SystemMessageIntent;
import com.projectswg.holocore.intents.support.objects.swg.DestroyObjectIntent;
import com.projectswg.holocore.intents.support.objects.swg.ObjectCreatedIntent;
import com.projectswg.holocore.resources.gameplay.world.travel.TravelGroup.ShuttleStatus;
import com.projectswg.holocore.resources.support.data.server_info.SdbLoader;
import com.projectswg.holocore.resources.support.data.server_info.SdbLoader.SdbResultSet;
import com.projectswg.holocore.resources.support.data.server_info.loader.DataLoader;
import com.projectswg.holocore.resources.support.data.server_info.mongodb.PswgDatabase;
import com.projectswg.holocore.resources.support.global.player.Player;
import com.projectswg.holocore.resources.support.objects.SpecificObject;
import com.projectswg.holocore.resources.support.objects.swg.SWGObject;
import com.projectswg.holocore.resources.support.objects.swg.creature.CreatureObject;
import me.joshlarson.jlcommon.concurrency.ThreadPool;
import me.joshlarson.jlcommon.log.Log;

import java.io.File;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Collection;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

public class TravelHelper {
	
	private final Map<String, TravelGroup> travel;
	private final ThreadPool travelExecutor;
	private final TravelPointManager pointManager;
	
	public TravelHelper() {
		this.travel = new ConcurrentHashMap<>();
		this.travelExecutor = new ThreadPool(3, "travel-shuttles-%d");
		this.pointManager = new TravelPointManager();
		
		createGalaxyTravels();
		loadTravelPoints();
	}
	
	public void start() {
		travelExecutor.start();
		for (TravelGroup gt : travel.values())
			travelExecutor.execute(gt);
	}
	
	public void stop() {
		travelExecutor.stop(true);
	}
	
	public void addTravelPoint(TravelPoint point) {
		pointManager.addTravelPoint(point);
	}
	
	public boolean isValidRoute(Terrain departure, Terrain destination) {
		return getTravelFee(departure, destination) != 0;
	}
	
	public int getTravelFee(Terrain departure, Terrain destination) {
		return DataLoader.Companion.travelCosts().getCost(departure, destination);
	}
	
	public TravelGroup getTravelGroup(String template) {
		return travel.get(template);
	}
	
	public List<TravelPoint> getAvailableTravelPoints(SWGObject traveller, Terrain terrain) {
		if (!isValidRoute(traveller.getTerrain(), terrain))
			return new ArrayList<>();
		TravelPoint nearest = getNearestTravelPoint(traveller);
		return pointManager.getPointsForTerrain(nearest, terrain);
	}
	
	public TravelPoint getDestinationPoint(SWGObject ticket) {
		Terrain planet = Terrain.getTerrainFromName(ticket.getAttribute("@obj_attr_n:travel_arrival_planet").split(":")[1]);
		String point = ticket.getAttribute("@obj_attr_n:travel_arrival_point");
		return getDestinationPoint(planet, point);
	}
	
	public TravelPoint getDestinationPoint(Terrain terrain, String pointName) {
		return pointManager.getDestination(terrain, pointName);
	}
	
	public TravelPoint getNearestTravelPoint(SWGObject object) {
		return pointManager.getNearestPoint(object.getWorldLocation());
	}
	
	public void grantTicket(TravelPoint departure, TravelPoint destination, SWGObject receiver) {
		// Create the ticket object
		SWGObject ticket = SpecificObject.SO_TRAVEL_TICKET.createType();
		Log.t("Granting ticket for departure: %s/%s  and destination: %s/%s", departure.getLocation(), departure.getName(), destination.getLocation(), destination.getName());
		
		// Departure attributes
		ticket.addAttribute("@obj_attr_n:travel_departure_planet", "@planet_n:" + departure.getTerrain().getName());
		ticket.addAttribute("@obj_attr_n:travel_departure_point", departure.getName());
		
		// Arrival attributes
		ticket.addAttribute("@obj_attr_n:travel_arrival_planet", "@planet_n:" + destination.getTerrain().getName());
		ticket.addAttribute("@obj_attr_n:travel_arrival_point", destination.getName());
		
		ticket.moveToContainer(receiver.getSlottedObject("inventory"));
		new ObjectCreatedIntent(ticket).broadcast();
	}
	
	public List<SWGObject> getTickets(CreatureObject creature) {
		Collection<SWGObject> tickets = creature.getItemsByTemplate("inventory", SpecificObject.SO_TRAVEL_TICKET.getTemplate());
		List<SWGObject> usableTickets = new ArrayList<>();
		
		for (SWGObject ticket : tickets) {
			if (isTicket(ticket) && isTicketUsable(ticket, getNearestTravelPoint(ticket)))
				usableTickets.add(ticket);
		}
		return usableTickets;
	}
	
	public boolean isTicket(SWGObject object) {
		String departurePlanet = object.getAttribute("@obj_attr_n:travel_departure_planet");
		String departureDestination = object.getAttribute("@obj_attr_n:travel_departure_point");
		String arrivalPlanet = object.getAttribute("@obj_attr_n:travel_arrival_planet");
		String arrivalPoint = object.getAttribute("@obj_attr_n:travel_arrival_point");
		
		return departurePlanet != null && departureDestination != null && arrivalPlanet != null && arrivalPoint != null;
	}
	
	public boolean isTicketUsable(SWGObject ticket, TravelPoint nearest) {
		String departurePoint = ticket.getAttribute("@obj_attr_n:travel_departure_point");
		String departurePlanet = ticket.getAttribute("@obj_attr_n:travel_departure_planet");
		Terrain departureTerrain = Terrain.getTerrainFromName(departurePlanet.split(":")[1]);
		Terrain currentTerrain = ticket.getTerrain();
		
		return departureTerrain == currentTerrain && departurePoint.equals(nearest.getName());
	}
	
	public void handleTicketUse(Player player, SWGObject ticket, TravelPoint nearestPoint, TravelPoint destinationPoint) {
		CreatureObject traveler = player.getCreatureObject();
		if (!isTicket(ticket)) {
			Log.e("%s attempted to use an object that isn't a ticket!", player);
		} else if (nearestPoint.getGroup().getStatus() != ShuttleStatus.GROUNDED) {
			int time = nearestPoint.getGroup().getTimeRemaining();
			new SystemMessageIntent(player, new ProsePackage(new StringId("travel/travel", "shuttle_board_delay"), "DI", time)).broadcast();
		} else if (!isTicketUsable(ticket, nearestPoint)) {
			// This ticket isn't valid for this point
			new SystemMessageIntent(player, "@travel:wrong_shuttle").broadcast();
		} else if (nearestPoint.isWithinRange(player.getCreatureObject())) {
			// They can use their ticket if they're within range.
			Log.i("%s/%s is traveling from %s to %s", player.getUsername(), traveler.getObjectName(), nearestPoint.getName(), destinationPoint);
			teleportAndDestroyTicket(destinationPoint, ticket, traveler);
		} else {
			new SystemMessageIntent(player, "@travel:boarding_too_far").broadcast();
		}
	}
	
	private void teleportAndDestroyTicket(TravelPoint destination, SWGObject ticket, CreatureObject traveler) {
		DestroyObjectIntent.broadcast(ticket);
		traveler.moveToContainer(destination.getCollector().getParent(), destination.getLocation());
	}
	
	private void createGalaxyTravels() {
		long groundTime = PswgDatabase.INSTANCE.getConfig().getInt(this, "shuttleGroundTime", 120);
		long airTime = PswgDatabase.INSTANCE.getConfig().getInt(this, "shuttleAirTime", 60);
		
		createGalaxyTravel(SpecificObject.SO_TRANSPORT_SHUTTLE.getTemplate(), 17, groundTime, airTime);
		createGalaxyTravel(SpecificObject.SO_TRANSPORT_STARPORT.getTemplate(), 21, groundTime, airTime);
		createGalaxyTravel(SpecificObject.SO_TRANSPORT_STARPORT_THEED.getTemplate(), 24, groundTime, airTime);
	}
	
	private void createGalaxyTravel(String template, long landTime, long groundTime, long airTime) {
		travel.put(template, new TravelGroup(landTime * 1000, groundTime * 1000, airTime * 1000));
	}
	
	private void loadTravelPoints() {
		try (SdbResultSet set = SdbLoader.load(new File("serverdata/travel/travel.sdb"))) {
			while (set.next()) {
				loadTravelPoint(set);
			}
		} catch (IOException e) {
			Log.e("Failed to load a travel point");
			Log.e(e);
		}
	}
	
	private void loadTravelPoint(SdbResultSet set) {
		String pointName = set.getText("name");
		double x = set.getReal("x");
		double z = set.getReal("z");
		String type = set.getText("type");
		Terrain travelPlanet = Terrain.getTerrainFromName(set.getText("planet"));
		if (travelPlanet == null) {
			Log.e("Invalid planet in travel.sdb: %s", set.getText("planet"));
			return;
		}
		
		TravelGroup group = getTravelGroupForType(type);
		TravelPoint point = new TravelPoint(pointName, new Location(x, 0, z, travelPlanet), group, type.endsWith("starport"));
		group.addTravelPoint(point);
		pointManager.addTravelPoint(point);
	}
	
	private TravelGroup getTravelGroupForType(String type) {
		switch (type) {
			case "shuttleport":
				return getTravelGroup(SpecificObject.SO_TRANSPORT_SHUTTLE.getTemplate());
			case "starport":
				return getTravelGroup(SpecificObject.SO_TRANSPORT_STARPORT.getTemplate());
			case "theed_starport":
				return getTravelGroup(SpecificObject.SO_TRANSPORT_STARPORT_THEED.getTemplate());
			default:
				Log.w("Invalid travel point type: %s", type);
				return null;
		}
	}
	
}

File: src/main/java/com/projectswg/holocore/resources/gameplay/world/travel/TravelHelper.kt
/***********************************************************************************
 * Copyright (c) 2025 /// Project SWG /// www.projectswg.com                       *
 *                                                                                 *
 * ProjectSWG is an emulation project for Star Wars Galaxies founded on            *
 * July 7th, 2011 after SOE announced the official shutdown of Star Wars Galaxies. *
 * Our goal is to create one or more emulators which will provide servers for      *
 * players to continue playing a game similar to the one they used to play.        *
 *                                                                                 *
 * This file is part of Holocore.                                                  *
 *                                                                                 *
 * --------------------------------------------------------------------------------*
 *                                                                                 *
 * Holocore is free software: you can redistribute it and/or modify                *
 * it under the terms of the GNU Affero General Public License as                  *
 * published by the Free Software Foundation, either version 3 of the              *
 * License, or (at your option) any later version.                                 *
 *                                                                                 *
 * Holocore is distributed in the hope that it will be useful,                     *
 * but WITHOUT ANY WARRANTY; without even the implied warranty of                  *
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the                   *
 * GNU Affero General Public License for more details.                             *
 *                                                                                 *
 * You should have received a copy of the GNU Affero General Public License        *
 * along with Holocore.  If not, see <http://www.gnu.org/licenses/>.               *
 ***********************************************************************************/
package com.projectswg.holocore.resources.gameplay.world.travel

import com.projectswg.common.data.encodables.oob.ProsePackage
import com.projectswg.common.data.encodables.oob.StringId
import com.projectswg.common.data.location.Location
import com.projectswg.common.data.location.Terrain
import com.projectswg.holocore.intents.support.global.chat.SystemMessageIntent
import com.projectswg.holocore.intents.support.objects.DestroyObjectIntent
import com.projectswg.holocore.intents.support.objects.ObjectCreatedIntent
import com.projectswg.holocore.resources.gameplay.world.travel.TravelGroup.ShuttleStatus
import com.projectswg.holocore.resources.support.data.server_info.SdbLoader
import com.projectswg.holocore.resources.support.data.server_info.SdbLoader.SdbResultSet
import com.projectswg.holocore.resources.support.data.server_info.loader.DataLoader.Companion.travelCosts
import com.projectswg.holocore.resources.support.data.server_info.mongodb.PswgDatabase.config
import com.projectswg.holocore.resources.support.global.player.Player
import com.projectswg.holocore.resources.support.objects.SpecificObject
import com.projectswg.holocore.resources.support.objects.swg.SWGObject
import com.projectswg.holocore.resources.support.objects.swg.creature.CreatureObject
import com.projectswg.holocore.resources.support.objects.swg.tangible.TangibleObject
import com.projectswg.holocore.resources.support.objects.swg.tangible.TicketInformation
import com.projectswg.holocore.utilities.HolocoreCoroutine
import com.projectswg.holocore.utilities.cancelAndWait
import me.joshlarson.jlcommon.log.Log
import java.io.File
import java.io.IOException
import java.util.concurrent.ConcurrentHashMap

class TravelHelper {
	private val travel: MutableMap<String, TravelGroup> = ConcurrentHashMap()
	private val coroutineScope = HolocoreCoroutine.childScope()
	private val pointContainer = TravelPointContainer()

	init {
		createGalaxyTravels()
		loadTravelPoints()
	}

	fun start() {
		for (group in travel.values) group.launch(coroutineScope)
	}

	fun stop() {
		coroutineScope.cancelAndWait()
	}

	fun addTravelPoint(point: TravelPoint) {
		pointContainer.addTravelPoint(point)
	}

	fun isValidRoute(departure: Terrain, destination: Terrain): Boolean {
		return getTravelFee(departure, destination) != 0
	}

	fun getTravelFee(departure: Terrain, destination: Terrain): Int {
		return travelCosts().getCost(departure, destination)
	}

	fun getTravelGroup(template: String): TravelGroup? {
		return travel[template]
	}

	fun getAvailableTravelPoints(traveller: SWGObject, terrain: Terrain): MutableList<TravelPoint> {
		if (!isValidRoute(traveller.terrain, terrain)) return ArrayList()
		val nearest = getNearestTravelPoint(traveller)
		return pointContainer.getPointsForTerrain(nearest, terrain)
	}

	fun getDestinationPoint(ticket: TangibleObject): TravelPoint? {
		val ticketInformation = ticket.ticketInformation
		val planet = ticketInformation.arrivalPlanet
		val point = ticketInformation.arrivalPoint
		return getDestinationPoint(planet, point)
	}

	fun getDestinationPoint(terrain: Terrain, pointName: String): TravelPoint? {
		return pointContainer.getDestination(terrain, pointName)
	}

	fun getNearestTravelPoint(obj: SWGObject): TravelPoint? {
		return pointContainer.getNearestPoint(obj.worldLocation)
	}

	fun grantTicket(departure: TravelPoint, destination: TravelPoint, receiver: SWGObject) {
		// Create the ticket object
		val ticket = SpecificObject.SO_TRAVEL_TICKET.createType() as TangibleObject
		Log.t("Granting ticket for departure: %s/%s  and destination: %s/%s", departure.location, departure.name, destination.location, destination.name)

		val ticketInformation = TicketInformation()
		// Departure attributes
		ticketInformation.departurePlanet = departure.terrain
		ticketInformation.departurePoint = departure.name


		// Arrival attributes
		ticketInformation.arrivalPlanet = destination.terrain
		ticketInformation.arrivalPoint = destination.name

		ticket.ticketInformation = ticketInformation
		ticket.moveToContainer(receiver.getSlottedObject("inventory"))
		ObjectCreatedIntent(ticket).broadcast()
	}

	fun getTickets(creature: CreatureObject): List<TangibleObject> {
		val tickets = creature.getItemsByTemplate("inventory", SpecificObject.SO_TRAVEL_TICKET.template)
		val usableTickets: MutableList<TangibleObject> = ArrayList()

		for (ticket in tickets) {
			if (ticket is TangibleObject) {
				if (isTicket(ticket) && isTicketUsable(ticket, getNearestTravelPoint(ticket))) usableTickets.add(ticket)
			}
		}
		return usableTickets
	}

	fun isTicket(obj: TangibleObject): Boolean {
		return obj.ticketInformation != null
	}

	fun isTicketUsable(ticket: TangibleObject, nearest: TravelPoint?): Boolean {
		val ticketInformation = ticket.ticketInformation
		val departurePoint = ticketInformation.departurePoint
		val departureTerrain = ticketInformation.departurePlanet
		val currentTerrain = ticket.terrain

		return departureTerrain == currentTerrain && departurePoint == nearest!!.name
	}

	fun handleTicketUse(player: Player, ticket: TangibleObject, nearestPoint: TravelPoint, destinationPoint: TravelPoint) {
		val traveler = player.creatureObject
		if (!isTicket(ticket)) {
			Log.e("%s attempted to use an object that isn't a ticket!", player)
		} else if (nearestPoint.group!!.status != ShuttleStatus.GROUNDED) {
			val time = nearestPoint.group.getTimeRemaining()
			SystemMessageIntent(player, ProsePackage(StringId("travel/travel", "shuttle_board_delay"), "DI", time)).broadcast()
		} else if (!isTicketUsable(ticket, nearestPoint)) {
			// This ticket isn't valid for this point
			SystemMessageIntent(player, "@travel:wrong_shuttle").broadcast()
		} else if (nearestPoint.isWithinRange(player.creatureObject)) {
			// They can use their ticket if they're within range.
			Log.i("%s/%s is traveling from %s to %s", player.username, traveler.objectName, nearestPoint.name, destinationPoint)
			teleportAndDestroyTicket(destinationPoint, ticket, traveler)
		} else {
			SystemMessageIntent(player, "@travel:boarding_too_far").broadcast()
		}
	}

	private fun teleportAndDestroyTicket(destination: TravelPoint, ticket: SWGObject, traveler: CreatureObject) {
		DestroyObjectIntent(ticket).broadcast()
		traveler.moveToContainer(destination.collector!!.parent, destination.location)
	}

	private fun createGalaxyTravels() {
		val groundTime = config.getInt(this, "shuttleGroundTime", 120).toLong()
		val airTime = config.getInt(this, "shuttleAirTime", 60).toLong()

		createGalaxyTravel(SpecificObject.SO_TRANSPORT_SHUTTLE.template, 17, groundTime, airTime)
		createGalaxyTravel(SpecificObject.SO_TRANSPORT_STARPORT.template, 21, groundTime, airTime)
		createGalaxyTravel(SpecificObject.SO_TRANSPORT_STARPORT_THEED.template, 24, groundTime, airTime)
	}

	private fun createGalaxyTravel(template: String, landTime: Long, groundTime: Long, airTime: Long) {
		travel[template] = TravelGroup(landTime + 10, groundTime, airTime)
	}

	private fun loadTravelPoints() {
		try {
			SdbLoader.load(File("serverdata/travel/travel.sdb")).use { set ->
				while (set.next()) {
					loadTravelPoint(set)
				}
			}
		} catch (e: IOException) {
			Log.e("Failed to load a travel point")
			Log.e(e)
		}
	}

	private fun loadTravelPoint(set: SdbResultSet) {
		val pointName = set.getText("name")
		val x = set.getReal("x")
		val z = set.getReal("z")
		val type = set.getText("type")
		val travelPlanet = Terrain.getTerrainFromName(set.getText("planet"))
		if (travelPlanet == null) {
			Log.e("Invalid planet in travel.sdb: %s", set.getText("planet"))
			return
		}

		val group = getTravelGroupForType(type)
		val point = TravelPoint(pointName, Location(x, 0.0, z, travelPlanet), group, type.endsWith("starport"))
		group!!.addTravelPoint(point)
		pointContainer.addTravelPoint(point)
	}

	private fun getTravelGroupForType(type: String): TravelGroup? {
		when (type) {
			"shuttleport"    -> return getTravelGroup(SpecificObject.SO_TRANSPORT_SHUTTLE.template)
			"starport"       -> return getTravelGroup(SpecificObject.SO_TRANSPORT_STARPORT.template)
			"theed_starport" -> return getTravelGroup(SpecificObject.SO_TRANSPORT_STARPORT_THEED.template)
			else             -> {
				Log.w("Invalid travel point type: %s", type)
				return null
			}
		}
	}
}
